tuna.go

```diff
@@ -777,21 +832,55 @@ func (c *Common) measureBandwidth(delayMeasuredSubs types.Nodes, n int) types.No
 				min, max, err := tunaUtil.BandwidthMeasurementClient(encryptedConn, int(c.MeasurementBytesDownLink), bandwidthTimeout)
 				dur := time.Since(timeStart)
 				if err != nil {
-					if _, ok := err.(net.Error); !ok {
-						log.Println(err)
+					if c.measureStorage != nil {
+						c.measureStorage.AddAvoidNode(sub.Metadata.Ip, &storage.AvoidNode{
+							IP:      sub.Metadata.Ip,
+							Address: sub.Address,
+						})
+						err = c.measureStorage.SaveAvoidNodes()
+						if err != nil {
+							log.Println(err)
+						}
+						log.Printf("add avoid node: %s", sub.Metadata.Ip)
 					}
 					return
 				}
 
 				log.Printf("address: %s, bandwidth: %f - %f KB/s, time: %s", addr, min/1000, max/1000, dur)
 
+				if c.measureStorage != nil {
+					var metadataString string
+					metadata, err := proto.Marshal(sub.Metadata)
+					if err != nil {
+						log.Println(err)
+						metadataString = ""
```

Comment:
You should not add favorite node if err != nil because the current logic assumes metadata is not empty